<!DOCTYPE html>
<html>
  <head>
    <title>Nolan Hughes - Blog</title>
    <!-- bootstrap -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"/>
    <!-- css file -->
    <link rel="stylesheet" type="text/css" href="../styles/main.css"/>
    <!-- Font api -->
    <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Sumana"/>
    <!-- favicon -->
    <link rel="icon" href="../img/icons/nh-logo-leftSideGray.png">
    <!-- jquery -->
    <script src="https://code.jquery.com/jquery-3.1.1.slim.js" integrity="sha256-5i/mQ300M779N2OVDrl16lbohwXNUdzL/R2aVUXyXWA=" crossorigin="anonymous"></script>
    <!-- js file -->
    <script src="../scripts/app.js"></script>
    <meta name="viewport" content="width=device-width,initial-scale=1">
  </head>

  <body>

    <div id="sidebar-wrapper" class="">
      <div id="close-btn" class="js-menu-close">
        <img src="../img/icons/x-blue.svg">
      </div>
      <div class="sidebar-links">
          <div class="accordion">
            <div class="section">
              <input type="checkbox" id="section-1"/>
              <label for="section-1"><span>Web Developer</span></label>
              <div id="list-close-btn" class="content">
                <ul>
                  <li><a href="../index.html">Home</a></li>
                  <li><a href="../index.html#bio">Bio</a></li>
                  <li><a href="../index.html#gallery">Projects</a></li>
                  <li><a href="../index.html#about">Experience</a></a></li>
                  <li><a href="../index.html#footer">Contact</a></li>
                </ul>
              </div>
            </div>

            <div class="section">
              <input type="checkbox" id="section-2"/>
              <label for="section-2"><span>Blog</span></label>
              <div id="list-close-btn" class="content">
                <ul>
                  <li><a href="../blog.html">Home</a></li>
                </ul>
              </div>
            </div>

          </div>
        </div>
      </div>

    <div id="main-wrapper">

      <div id="intro-blog-bg">
        <div id="intro" class="container-fluid hero landscape blog-intro">
          <div class="logo col-md-6">
            <a href="../blog.html"><img src="../img/icons/nh-logo-leftSideGray.png" /></a>
          </div>
          <div id="hamburger" class="js-menu-open col-md-6">
            <img src="../img/icons/hamburgerButton-gray.png"/>
          </div>
          <h1>Promises</h1>
          <h3>Part 2</h3>
          <h4 class="post-sub-heading">Posted by Nolan Hughes on November 5th, 2018</h4>
        </div>
      </div>

      <div id="body" class="container-fluid hero md-box">
        <div class="inside-container-blog inside-container">
          <div class="posts">
            <p>
              For part 2 of my <code class="inline-code">Promises</code> series I thought it’d be a good idea to go over error propagation. The main takeaways are using <code class="inline-code">.catch</code> and try and catch with <code class="inline-code">aysnc/await</code>. Using <code class="inline-code">.catch</code> at the end of a promise chain looks like the example below.
            </p>

            <code>
              <span>doSomething()</span>
              <span>.then(result => doSomethingElse(result))</span>
              <span>.then(newResult => doThirdThing(newResult))</span>
              <span>.then(finalResult => console.log(`Got the final result: ${finalResult}`))</span>
              <span>.catch(failureCallback);</span>
            </code>

            <p>
              When you do this the chain will stop if there’s an exception because it then looks down the chain for a catch handler. An even nicer and more succinct way is to use <code class="inline-code">aysnc/await</code> with try and catch. This example is very reminiscent of synchronous code.
            </p>

            <code>
              <span>async function foo() {</span>
              <span style="text-indent: 23px">try {</span>
              <span style="text-indent: 46px">  const result = await doSomething();</span>
              <span style="text-indent: 46px">  const newResult = await doSomethingElse(result);</span>
              <span style="text-indent: 46px">  const finalResult = await doThirdThing(newResult);</span>
              <span style="text-indent: 46px">  console.log(`Got the final result: ${finalResult}`);</span>
              <span style="text-indent: 23px">} catch(error) {</span>
              <span style="text-indent: 46px">  failureCallback(error);</span>
              <span style="text-indent: 23px">}</span>
              <span>}</span>
            </code>

            <p>
              This is wrapped up in a super neat <code class="inline-code">async</code> function and catches all of your errors. Even thrown exceptions and programming errors! Being able to handle your errors is an essential part of writing clean code and for functional composition of asynchronous operations.            
            </p>

          </div>
        </div>
      </div>

      <div id="footer">
        <div class="inside-container row footer-content">
          <div class="right-column" id="blog-footer">
            <ul>
              <li>
                <a href="https://learn.co/NolanHughes" target="_blank">
                  <img src="../img/icons/Flatiron-logo.png"/>
                </a>
              </li>
              <li>
                <a href="https://github.com/NolanHughes" target="_blank">
                  <img src="../img/icons/github-logo-3.png"/>
                </a>
              </li>
            </ul>
            <p>©2017 Nolan Hughes</p>
          </div>
        </div>
      </div>
    </div>
  </body>
</html>
